timemachine/cpp/src/local_config.cu

```diff
@@ -0,0 +1,190 @@
+    // to provide a new seed at each step.
+    void LocalMDConfig::setup_from_idxs(
+        double *d_x_t, double *d_box_t, const std::vector<int> &local_idxs, const double temperature, const int seed, const double radius, const double k, const cudaStream_t stream) {
+
+        curandErrchk(curandSetStream(cr_rng_, stream));
+        curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+
+        // Set the array to all N, which indicates to ignore that idx
+        k_initialize_array<unsigned int>
+            <<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(N_, d_free_indices_.data, N_);
+        gpuErrchk(cudaPeekAtLastError());
+
+
+        // Generate values between (0, 1.0]
+        curandErrchk(curandGenerateUniform(cr_rng_, probability_buffer_.data, round_up_even(N_)));
+
+        std::mt19937 rng;
+        rng.seed(seed);
+        std::uniform_int_distribution<unsigned int> random_dist(0, local_idxs.size() - 1);
+
+
+        unsigned int reference_idx = local_idxs[random_dist(rng)];
+
+        const double kBT = BOLTZ * temperature;
+        // Select all of the particles that will be free
+        k_log_probability_selection<float><<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(
+            N_, kBT, radius, k, reference_idx, d_x_t, d_box_t, probability_buffer_.data, d_free_indices_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
+        this->_setup_free(k, radius, reference_idx, stream);
+    }
+
+    void LocalMDConfig::_setup_free(const double k, const double radius, const unsigned int reference_idx, const cudaStream_t stream) {
+        const int tpb = warp_size;
+
+        LessThan select_op(N_);
+
+        // Partition the free indices into the row indices
+        gpuErrchk(cub::DevicePartition::If(
+            d_temp_storage_buffer_->data,
+            temp_storage_bytes_,
+            d_free_indices_.data,
+            d_row_idxs_.data,
+            num_selected_buffer_.data,
+            N_,
+            select_op,
+            stream));
+
+        gpuErrchk(cudaMemcpyAsync(
+            p_num_selected_.data,
+            num_selected_buffer_.data,
+            1 * sizeof(*p_num_selected_.data),
+            cudaMemcpyDeviceToHost,
+            stream));
+        gpuErrchk(cudaStreamSynchronize(stream));
+
+        const int num_row_indices = p_num_selected_.data[0];
+        const int num_col_indices = N_ - num_row_indices;
+
+        if (num_row_indices == 0) {
+            throw std::runtime_error("LocalMDConfig setup has no free particles selected");
+        }
+
+        // The reference particle will always be in the column indices
+        if (num_row_indices == N_ - 1) {
+            fprintf(stderr, "LocalMDConfig setup has entire system selected\n");
+        }
+
+        k_construct_bonded_params<<<ceil_divide(num_row_indices, tpb), tpb, 0, stream>>>(
+            num_row_indices, N_, reference_idx, k, 0.0, radius, d_row_idxs_.data, restraints_.data, bond_params_.data);
```

Comment:
From looking at this together offline with @badisa , I think this may have been part of the explanation for a discrepancy relative to `master`.

(If random # of restraints in current iteration is higher than random # of restraints in previous iteration, then these may not be zeroed out in `restraints_`, `bond_params_` due to this check https://github.com/proteneer/timemachine/blob/84442e30b4f62ea305722f969b6a983690d282e2/timemachine/cpp/src/kernels/k_local_md.cuh#L15-L17  .

Previously, `k_construct_bonded_params` was only invoked once immediately after construction, rather than repeatedly on the same `restraints_`, `bond_params_` arrays, which can now contain stale non-zero values at `idxs` > current `num_idxs`.)